Develop pc bv1.0 mme301

nonphixi0n opened the pull request.

nonphixi0n committed:
Добавлена таблица коммутации. Расписаны индексы реле.

nonphixi0n committed:
Исправлена логика коммутация основных режимов измерений.

nonphixi0n committed:
Добавил новые файлы, относящиеся к предыдущему коммиту.

nonphixi0n committed:
Добавлен регистр выбора позиции имзерения.

nonphixi0n committed:
Добавлен обработчик прерываний линии системы безопасности. Добавлена команда на деактивацию контура юезопасности.

nonphixi0n committed:
Добавлена логика режима самотестирования.

nonphixi0n committed:
Добавлена логика среднего уровня

nonphixi0n committed:
Доработки альфа версии

asytyi commented on Firmware/Source/Controller/Commutator.c:
```diff
@@ -0,0 +1,491 @@
+	{
+		case ACT_COMM_ILEAK_GATE_EMITTER_POS_PULSE:
+			{
+				if (REG_MEASUREMENT_POSITION == 1)
```
Здесь и далее вместо
`if (REG_MEASUREMENT_POSITION == 1)`
надо
`if (DataTable[REG_MEASUREMENT_POSITION] == 1)`

asytyi commented on Firmware/Source/Controller/DebugActions.c:
```diff
@@ -10,8 +10,12 @@
+uint8_t SPI_Data_Ileak_Gate_Emitter_Pos[] = {0xFF, 0xA3, 0xCE, 0xBC, 0x3F};
+uint8_t SPI_Data_Ileak_Gate_Emitter_Neg[] = {0xFF, 0xA3, 0xAF, 0x3C, 0x3F};
+uint8_t SPI_Debug[] = {0x00, 0x00, 0x01, 0x00, 0x00};
+uint8_t Data_Length = 4;
```
Выбранный способ формирования отладочных функций является неудачным. В итоге эти функции дублируют таблицу коммутации. Например, при изменении платы надо помнить, что потребуется внесение изменений в двух местах. При этом сами отладочные функции лишены гибкости.
Будет правильнее сделать, например, две отладочные функции которые умеют следующее:

1. Функция, которая считывает из отладочного регистра номер из таблицы коммутации. И для этого номера можно подать или снять высокий уровень (включить/выключить реле)
2. Функция, которая считывает из отладочного регистра номер регистра и маску. И записывает в сдвиговый регистр с соответствующим номером заданную маску.

Это гибкий подход, который позволит при необходимости пошагово всё проверить.

asytyi replied:
Не устранено, но если для отладки не требуется, то не настаиваю

asytyi commented on Firmware/Source/Controller/DebugActions.c:
```diff
@@ -49,25 +53,113 @@ void DBACT_WriteSPI()
+{
+	LL_SPITurnOffSS();
+}
+//-----------------------
```
Такие обёртки не несут особого смысла и только перегружают код. Почему нельзя просто использовать `LL_SPITurnOnOE` и какую ценность даёт `DBACT_SPITurnOnOE`? Именно здесь посоветую не загружать код обёрточными функциями и их объявлениями.

При этом глобально я обёртки ценю и уважаю, но когда они несут смысл. Например, если есть функция `SetPort1(true)`  оборачивается в какую-нибудь `TurnOnIndication()`. Вопросов нет, читаемость улучшилась.

asytyi commented on Firmware/Source/Controller/Global.h:
```diff
@@ -13,10 +13,12 @@
 #define TIME_LED_BLINK							500		// Мигание светодиодом (в мс)
 #define TIME_FAULT_LED_BLINK					250		// Мигание светодиодом в состоянии Fault
+#define COMM_DELAY_NONE_MS						100
+#define COMM_DELAY_LONG_MS						500
```
Копипасту убрать, задержку поправить до адекватного значения

asytyi commented on Firmware/Source/Platform/DeviceObjectDictionary.h:
```diff
@@ -2,34 +2,62 @@
 // Сохраняемые регистры
 #define REG_SFTST_V_ALLOWED_ERR			0	// Допустимая погрешность измеренного значения напряжения (%)
+#define REG_PS_FIRST_START_TIME			1	// Время готовности при первом старте
+#define REG_SELF_TEST_ACTIVE			2	// Флаг состояния системы самотестирования
```
Почему этот регистр находится в области NVRAM?

asytyi commented on Firmware/Source/Controller/SelfTest.c:
```diff
@@ -4,13 +4,225 @@
+			break;
+
+		case SS_ST_InputRelayCheck_1:
+			ST_State = 1;
```
Абсолютно лишняя сущность. Объяви в функции переменную
`static DeviceSubState PrevSubstate = SS_None`
а в кейсах включения коммутации обновляешь её, например
`PrevSubstate  = SS_ST_InputRelayCheck_1`

asytyi commented on Firmware/Source/Controller/LowLevel.c:
```diff
@@ -90,6 +139,7 @@ bool LL_SelfTestMeasure()
-	Error = (MeasuredTestVoltage - ADC_V_CC) / ADC_V_CC * 100;
-	return (Error >= DataTable[REG_SFTST_V_ALLOWED_ERR]) ? false : true;
+	// Error = (MeasuredTestVoltage - ADC_V_CC) / ADC_V_CC * 100;
+	// return (Error >= DataTable[REG_SFTST_V_ALLOWED_ERR]) ? false : true;
```
Закомментированный код коммитить запрещено

asytyi commented on Firmware/Source/Controller/Logic.c:
```diff
@@ -28,3 +29,13 @@ void LOGIC_ResetOutputRegisters()
+
+void LOGIC_UpdatePosition()
+{
+	MeasurementPosition = DataTable[REG_MEASUREMENT_POSITION];
```
Ето что и для чего? Переменная нигде не используется. Вообще можно прибить эту либу, а функции перенести в Controller, если нужны

asytyi commented on Firmware/Source/Controller/Controller.c:
```diff
@@ -120,14 +124,58 @@ static Boolean CONTROL_DispatchAction(Int16U ActionID, pInt16U pUserError)
+}
+//-----------------------------------------------
+
+bool CONTROL_Delay(Int16U Time)
+{
+	static Int64U CONTROL_Delay = 0;
+
+	if(!CONTROL_Delay)
+		CONTROL_Delay = CONTROL_TimeCounter + Time;
+	else
+	{
+		if(CONTROL_TimeCounter >= CONTROL_Delay)
+		{
+			CONTROL_Delay = 0;
+			return true;
+		}
+	}
+
+	return false;
```
Неудачная реализация функции. Нормально отработает, если будет вызвана по истечению таймаута. Если цикл проверки по каким-то причинам будет прерван, то следующее выставление таймаута будет косячным. Именно здесь в коде возможно такого не будет. Но не факт, что в дальнейшем не появится такой функционал, или реализация будет скопирована в проект, где это важно. Поэтому сразу надо делать правильно. А правильно вот так:

В `void CONTROL_LogicProcess()` объявляешь переменную
`static Int64U Delay = 0`
и в свиче инициализацию и проверку задержки делаешь в отдельных кейсах, т.е.
```
switch(CONTROL_SubState)
{
	case SS_StartUp:
		Delay = CONTROL_TimeCounter + DataTable[REG_PS_FIRST_START_TIME]
		CONTROL_SetDeviceState(DS_InProcess, SS_WaitDelay);
		break;
	
	case SS_WaitDelay:
		if(CONTROL_TimeCounter > Delay)
			....
		break;
```

Будет работать железобетонно в любых случаях.

asytyi commented on Firmware/Source/Controller/Controller.c:
```diff
@@ -120,14 +124,58 @@ static Boolean CONTROL_DispatchAction(Int16U ActionID, pInt16U pUserError)
+			case SS_StartSelfTest:
+				if(DataTable[REG_SELF_TEST_ACTIVE])
+				{
+					CONTROL_SetDeviceState(DS_SelfTest, SS_ST_StartPrepare);
+				}
+				else
+					CONTROL_SetDeviceState(DS_InProcess, SS_Commutation);
```
Как будто `DS_SelfTest` и `DS_InProcess` дублируют друг друга.
Непосредственно включение коммутации для измерения прибора является синхронным процессом. Т.е. поступает команда на включение коммутации (я имею ввиду команды 100-110) и она быстро исполняется в самом обработчике.
Длительным процессом является только селф тест. Ну вот надо оставить из девайс стейтов только 
`DS_InProcess` в рамках которого обрабатываются сабстейты.

asytyi commented on Firmware/Source/Controller/Commutator.c:
```diff
@@ -0,0 +1,491 @@
+}
+// ----------------------------------------
+
+void COMM_Commutate(Int16U ActionID)
```
Функция нигде не вызывается

nonphixi0n committed:
Исправлено условие выбора позиции

nonphixi0n committed:
Исправлена задержка коммутации

nonphixi0n committed:
Исправлен регистр % ошибки оцифровки

nonphixi0n committed:
1. Исправлена отладочная функция SPI; 2. Удалены лишние функции нижнего уровня; 3. Исправлена функция задержки; 4. Исправлена логика библиотеки SelfTest.c; 5. Убраны лишние DeviceState, 5. Добавлен вызов функции коммутации COMM_Commutate.

asytyi commented on Firmware/Source/Platform/DeviceObjectDictionary.h:
```diff
@@ -51,13 +41,12 @@
 // Несохраняемые регистры чтения-записи
 #define REG_MEASUREMENT_POSITION		128	// Регистр выбора позции при измерении (1 или 2)
+#define REG_SFTST_V_ALLOWED_ERR			129	// Допустимая погрешность измеренного значения напряжения (%)
+#define REG_SELF_TEST_ACTIVE			130	// Флаг состояния системы самотестирования
```
Если `REG_SELF_TEST_ACTIVE` статусный регистр, то его надо поместить в read-only область, т.е. регистры, которые начинаются от 192. 192-197 стандартные регистры состояния, их не трогать)
А вот `REG_SFTST_V_ALLOWED_ERR` зря перенёс. Уровень ошибки будет единожды настраиваться при запуске блока, и ВУ ничего не будет в этом плане задавать. Поэтому место ему в NVRAM области, где он и был

asytyi replied:
Кстати, вместо регистра REG_SFTST_V_ALLOWED_ERR  в коде в SelfTest.c используется дефайн. Нужно всё-таки регистр

nonphixi0n replied:
Да, флаг режима самотестирования статусный. Перенес на 200. Регистр допустимой погрешности вернул, и в SelfTest.c убрал дефайн и выразил через DataTable.

asytyi commented on Firmware/Source/Platform/DeviceObjectDictionary.h:
```diff
@@ -51,13 +41,12 @@
 
 // Регистры
 // Сохраняемые регистры
-#define REG_SFTST_V_ALLOWED_ERR			0	// Допустимая погрешность измеренного значения напряжения (%)
```
Надо выровнять табуляции, чтобы все числовые значения шли в один ряд

nonphixi0n replied:
Поправил.

asytyi commented on Firmware/Source/Controller/ZcRegistersDriver.h:
```diff
@@ -13,6 +13,7 @@
 //
 // Reset output register to zero
 void ZcRD_RegisterReset();
+void ZcRD_OutputValuesReset1();
```
Это что за чудо?

nonphixi0n replied:
Блин, вчерашнее наследие от отладки. Убрал.

asytyi commented on Firmware/Source/Platform/DeviceObjectDictionary.h:
```diff
@@ -11,20 +11,10 @@
+
 
 #define ACT_COMM_NONE							100 // Отключение всех реле
 #define ACT_COMM_ILEAK_GATE_EMITTER_POS_PULSE	101 // Режим измерения тока утечки затвора-эмиттера (положительный импульс)
```
Сейчас обратил внимание. Заряд затвора измеряется при единственной коммутации. Поэтому `ACT_COMM_Q_GATE_NEG_PULSE` надо выпилить, а `ACT_COMM_Q_GATE_POS_PULSE` переименовать в `ACT_COMM_Q_GATE`

nonphixi0n replied:
Удалил лишнюю коммутацию.

asytyi commented on Firmware/Source/Controller/ZcRegistersDriver.c:
```diff
@@ -38,21 +38,44 @@ void ZcRD_OutputValuesCompose(Int16U TableID, Boolean TurnOn)
+	for (int i = 0; i <= Data_Lenght; i++)
+	{
+		SPI_WriteByte8b(SPI1, CurrentOutputValuesCopy[i]);
+		//DELAY_MS(1);
```
Закомментированный код убрать

nonphixi0n replied:
Исправил.

asytyi commented on Firmware/Source/Controller/ZcRegistersDriver.c:
```diff
@@ -38,21 +38,44 @@ void ZcRD_OutputValuesCompose(Int16U TableID, Boolean TurnOn)
 
-	for (i = 0; i < COMMUTATION_EXT_BOARDS; ++i)
-		CurrentOutputValues[i] = 0;
+void ZcRD_OutputValuesReset1()
```
Мёртвый код убрать

asytyi commented on Firmware/Source/Controller/ZcRegistersDriver.c:
```diff
@@ -38,21 +38,44 @@ void ZcRD_OutputValuesCompose(Int16U TableID, Boolean TurnOn)
-	Int16U CurrentOutputValuesCopy[COMMUTATION_EXT_BOARDS];
+	int i;
+	uint8_t CurrentOutputValuesCopy[14];
+	uint8_t Data_Lenght = 14;
```
Магических чисел быть не должно. 14 вынести в осмысленный дефайн. Лучше всего сделать отдельные дефайны по числу регистров на каждой плате, например
```
#define NUM_REGS_INPUT_BOARD 1
#define NUM_REGS_HV_BOARD 3
```
а затем дефайн общего числа регситров и его использовать
```
#define NUM_REGS_TOTAL    (NUM_REGS_INPUT_BOARD + NUM_REGS_HV_BOARD)
```
Читаемость кода сильно улучшится

nonphixi0n replied:
Добавил дефайны.

asytyi commented on Firmware/Source/Controller/SelfTest.c:
```diff
@@ -22,17 +21,17 @@
 		case SS_ST_StartPrepare:
-			ST_State = 0;
-			CONTROL_SetDeviceState(DS_SelfTest, SS_ST_InputRelayCheck_1);
+			PrevSubstate = SS_None;
```
Здесь и далее исправить присвоение PrevSubstate как описано в треде в слэке

nonphixi0n replied:
Сделал.

asytyi commented on Firmware/Source/Controller/SelfTest.c:
```diff
@@ -166,7 +165,7 @@ void SELFTEST_Process()
+			CONTROL_SetDeviceState(DS_InProcess, SS_ST_CurrentMeasure);
 			break;
 
 		case SS_ST_CurrentMeasure:
```
В операторе
```
if (Error >= ST_V_ALLOWED_ERR_DEF)
{
	DataTable[REG_SELF_TEST_OP_RESULT] = OPRESULT_FAIL;
	CONTROL_SwitchToFault(DF_VOLTAGE_MEASURING);
}
```
добавь статусный регистр в который будет записываться значение PrevSubstate. Это позволит определять на каком шаге произошёл фэйл.

nonphixi0n replied:
Добавил регистр.

asytyi commented on Firmware/Source/Controller/LowLevel.c:
```diff
@@ -32,84 +32,32 @@ void LL_SetStateSFGreenLed(bool State)
 {
 	GPIO_SetState(GPIO_SPI_RST, false);
-	DELAY_MS(10);
+	DELAY_MS(1);
```
Не велика ли задержка в 1мс для сброса? Как будто достаточно в тысячу раз меньше `DELAY_US(1);`

nonphixi0n replied:
Исправил.

asytyi commented on Firmware/Source/Controller/Global.h:
```diff
@@ -13,12 +13,11 @@
 #define ADC_REF_VOLTAGE							3300.0f	// Опорное напряжение (в В)
 #define ADC_RESOLUTION							4095	// Разрешение АЦП
-#define ADC_V_CC								0.1f	// Напряжение при замкнутой цепи (в В)
+#define ADC_V_CC								150.0f	// Напряжение при замкнутой цепи (в В)
```
Напряжение в вольтах?

nonphixi0n replied:
В мв.

asytyi replied:
Проверка связи. Уведомления об ответах приходят?

asytyi replied:
@nonphixi0n и вот так

asytyi commented on Firmware/Source/Controller/Controller.c:
```diff
@@ -111,107 +141,141 @@ static Boolean CONTROL_DispatchAction(Int16U ActionID, pInt16U pUserError)
+			case SS_ST_MCRelayCheck_3:
+			case SS_ST_MCRelayCheck_4:
+			case SS_ST_CurrentMeasure:
+				break;
```
Вместо этого достаточно написать
```
    default:
        break;
```

asytyi commented on Firmware/Source/Controller/Controller.c:
```diff
@@ -111,107 +141,141 @@ static Boolean CONTROL_DispatchAction(Int16U ActionID, pInt16U pUserError)
+				{
+					CONTROL_TimeCounter = CONTROL_TimeCounter + 1;
+					CONTROL_SetDeviceState(DS_InProcess, SS_WaitDelay);
+				}
```
Здесь происходит что-то непонятное. `CONTROL_TimeCounter` уже объявлена и является глобальным счётчиком, который инкрементируется в прерывании. Зачем она здесь как статик объявляется? Код второго кейса должен выглядеть вот так
```
			case SS_WaitDelay:
				if (CONTROL_TimeCounter >= Delay)
					CONTROL_SetDeviceState(DS_InProcess, SS_WaitDelay);
				break;
```

nonphixi0n replied:
Здесь немного не понял. Сейчас сделал так, что при достижении счетчиком значения больше или равным задержке, сабстейт переходит в режим самотестирования.

asytyi reviewed (changes requested):
Есть замечания

nonphixi0n committed:
Поправил расположение регистров в памяти. Теперь в SelfTest.c погрешность считывается через регистр.

nonphixi0n committed:
Выровнил табуляцию в DeviceObjectDictionary.h. Убрал лишние функции из ZcRegistersDriver.h. Удалил коммутацию ACT_COMM_Q_GATE_NEG_PULSE.

nonphixi0n committed:
Убрал лишний код, добавил дефайны в ZcRegistersDriver.c, исправил присвоение PrevSubstate/

nonphixi0n committed:
Исправлена длительность задержки в функциях нижнего уровня. Поправил задержку в CONTROL_LogicProcess.

nonphixi0n committed:
Переработана таблица коммутации (HV-реле состоит из 4 шт.)

asytyi commented on Firmware/Source/Controller/Commutator.c:
```diff
@@ -29,6 +30,61 @@ void COMM_CommutateNone()
 }
 // ----------------------------------------
 
+void COMM_PE()
```
Функцию лучше назвать `COMM_DisconnectPE()`

nonphixi0n replied:
Понял. Переименую

nonphixi0n replied:
Готово

nonphixi0n committed:
Функция COMM_PE переименована в COMM_DisconnectPE

asytyi commented on Firmware/Source/Controller/Commutator.h:
```diff
@@ -13,6 +13,8 @@
 // Switch commutation to none
 void COMM_CommutateNone();
+// Switch PE
+void COMM_PE();
```
`COMM_DisconnectPE()` используется в пределах библиотеки, поэтому внешнее объявление лишнее. Именно поэтому оно осталось незамеченным после правок

nonphixi0n replied:
Сделано

asytyi replied:
Стоп. Зачем ты переименовал `COMM_DisconnectPE()` в `COMM_СonnectPE()`? Разве эта функция включает землю? Она же подаёт высокий уровень на NC реле, что приводит отключению земли? Или я что-то неправильно понимаю?

nonphixi0n replied:
Да, точно. Немного запутался. Переименовал обратно.

asytyi commented on Firmware/Source/Controller/Commutator.c:
```diff
@@ -46,56 +103,36 @@ void COMM_Commutate(Int16U ActionID)
 				{
 					ZcRD_OutputValuesReset();
 
+					COMM_DisconnectPE();
```
`ZcRD_OutputValuesReset()` используется и до вызова `COMM_DisconnectPE()` и в нём самом. Такое неявное поведение лучше избегать. Лучше в `COMM_DisconnectPE()` вызов удалить и оставить только функции `ZcRD_OutputValuesCompose`

nonphixi0n replied:
Понятно, исправил

asytyi commented on Firmware/Source/Controller/DebugActions.c:
```diff
@@ -76,9 +78,10 @@ void DBACT_SelfTestMeasure()
 	Error = (Result - ADC_V_CC) / Result * 100;
 
-	if (Error >= 20)
+	if (Error <= 20)
```
Во-первых, этот код используется и в DebugActions, и в SelfTest, но с отличиями. Надо сделать одну функцию и её вызывать в двух местах, а не копипастить код.
Во-вторых, сам код какой-то замысловатый. Есть какая-то константа 150мВ, есть ошибка. Зачем так? Если известно падение напряжения при нормальном функционировании, то просто задай порог ошибочного напряжения, хранимого в регистре, и с ним сравнивай.
Например
```
bool TestVoltageFailed()
{
	return LL_SelfTestMeasure() > DataTable[REG_SFTST_V_ALLOWED_VOLTAGE];
}
```
просто и без излишеств.
Если нормальное напряжение 150мВ, то пусть в регистре будет 175-200мВ, например. Не знаю какой там разброс результатов.

nonphixi0n replied:
Готово

asytyi commented on Firmware/Source/Controller/Controller.c:
```diff
@@ -157,8 +157,8 @@ bool CONTROL_DispatchAction(Int16U ActionID, pInt16U pUserError)
 				*pUserError = ERR_OPERATION_BLOCKED;
-			else if(CONTROL_State == DS_None)
-				*pUserError = ERR_DEVICE_NOT_READY;
+			//else if(CONTROL_State == DS_None)
+				//*pUserError = ERR_DEVICE_NOT_READY;
```
Ето что такое?

asytyi replied:
Выше в `CONTROL_Idle` тоже закомментированный код. Дима, возьми за правило во время коммитов делать правый клик по коммитящимся файлам и просматривай изменения через Show Diff.
**Для всех изменённых файлов. Всегда.**
Чтобы не коммитить всякую непотребу

nonphixi0n replied:
Понял. Код исправил

asytyi commented on Firmware/Source/Controller/Global.h:
```diff
@@ -16,8 +16,8 @@
 #define COMM_DELAY_MS							1
```
По документации время переключения реле 8мс (фактическое поменьше). Почему ждём всего 1мс?

nonphixi0n replied:
Исправил

asytyi commented on Firmware/Source/Controller/SelfTest.c:
```diff
@@ -169,13 +168,18 @@ void SELFTEST_Process()
+				DataTable[REG_FAULT_RELAY] = PrevSubstate;
+				ZcRD_RegisterReset();
 				CONTROL_SwitchToFault(DF_VOLTAGE_MEASURING);
 			}
```
Как я понимаю, процедура самотестирования в данный момент умеет только определять, что реле включаются и падение напряжение на них ниже определённого уровня. Но не умеет определять залипания.
Код не дорабатывать, обсудить в слэке.

asytyi commented on Firmware/Source/Controller/ZcRegistersDriver.c:
```diff
@@ -43,36 +52,32 @@ void ZcRD_OutputValuesReset()
-		CurrentOutputValues[i] = 0x00;
-}
-// ----------------------------------------
+	for (i = 0; i <= NUM_REGS_TOTAL; ++i)
+		CurrentOutputValues[i] = 0;
```
Вылетел за границы массива. Если массив размера `NUM_REGS_TOTAL`, то проход по всем элементам
`for (i = 0; i < NUM_REGS_TOTAL; ++i)`

nonphixi0n replied:
Исправил

asytyi commented on Firmware/Source/Controller/ZcRegistersDriver.c:
```diff
@@ -43,36 +52,32 @@ void ZcRD_OutputValuesReset()
 	GPIO_SetState(GPIO_SPI_OE, true);
 
-	for (int i = 0; i <= Data_Lenght; i++)
+	for (int i = 0; i <= (NUM_REGS_TOTAL - 1); i++)
```
Зачем создаётся копия массива?
- Придерживайся единого стиля объявления индексной переменной. У тебя она и в функции объявляется `int i;`, и потом локально в итерраторе массива `for (int i = 0; i <= (NUM_REGS_TOTAL - 1); i++)`
- В функциях выше для индекса используется `uint8_t`, здесь `int`, хотя массивы одинакового размера
- Лучше объявлять индекс в итераторе, а для прохода по всему массиву использовать  такую запись
`for (uint8_t i = 0; i < NUM_REGS_TOTAL; i++)`

Это всё мелочи, но из-за этих мелочей код выглядит непрофессионально и трудночитаемо

nonphixi0n replied:
Сделано

nonphixi0n replied:
Копию массива сделал для реверса самого массива

asytyi replied:
Объясни момент с реверсом. Может я что-то недопонял в коде

nonphixi0n replied:
Потому что, у меня CommutationTable составлена в таком порядке: OutHVRealayBoard_1, OutHVRealayBoard_2, OutThermRelayBoard, InputRelayBoard. А подключение шлейфами в обратном: InputRelayBoard, OutThermRelayBoard, OutHVRealayBoard_2, OutHVRealayBoard_1.

asytyi commented on Firmware/Source/Platform/DeviceObjectDictionary.h:
```diff
@@ -41,52 +39,55 @@
-#define REG_SELF_TEST_ACTIVE			130	// Флаг состояния системы самотестирования
-#define REG_DBG							150	// Отладочный регистр
+#define REG_MEASUREMENT_POSITION				128	// Регистр выбора позции при измерении (1 или 2)
+#define REG_FAULT_RELAY							129	// Faulty Relay
```
Этот регистр ведь только для чтения для верхнего уровня. Так почему он в записываемой области?

nonphixi0n replied:
Саш, какой регистр ты имеешь в виду?

asytyi reviewed (changes requested):
Необходимо устранить замечания

nonphixi0n committed:
Изменил название функции COMM_DisconnectPE() на COMM_ConnectPE()

nonphixi0n committed:
Внесены исправления в библиотеку ZcRegistersDriver.c

asytyi commented on Firmware/Source/Controller/Controller.c:
```diff
@@ -110,60 +141,119 @@ static Boolean CONTROL_DispatchAction(Int16U ActionID, pInt16U pUserError)
+					CONTROL_SetDeviceState(DS_InProcess, SS_InProcess);
+				break;
+
+			case SS_InProcess:
+					if(REG_SELF_TEST_OP_RESULT == 1)
+					{
+						CONTROL_SetDeviceState(DS_InProcess, SS_InProcess);
+					}
+					else
+						CONTROL_SetDeviceState(DS_Fault, SS_None);
+				break;
+
+			default:
+				break;
```
Здесь совсем ерунда:
1. `SS_InProcess` и `DS_InProcess` выглядят одинаково. Назови `SS_InProcess` каким-то более говорящим названием
2. `if(REG_SELF_TEST_OP_RESULT == 1)` сравнение не со значением регистра, а с его номером
3. Даже если исправить сравнение на значение регистра, то получается, если оно равно 1 (Кстати, 1 это что такое? Почему используется число, а не привязанный к нему дефайн?), то опять переходим к `CONTROL_SetDeviceState(DS_InProcess, SS_InProcess);`. Это же зацикливание.

nonphixi0n replied:
1. Исправил наименование
2. Поправил обращение к регистру
3. Действительно, изменил переход на нужные DeviceState и SubState

nonphixi0n committed:
Привел проверку замкнутых реле в фалах DebugActions.c и SelfTest.c к одному виду

nonphixi0n committed:
Исправил цикличность в КА CONTROL_LogicProcess()

nonphixi0n committed:
Добавлен код для двух цепей самотестирования залипших реле

nonphixi0n committed:
Исправил ошибку в условии цикла

nonphixi0n committed:
Добавил 2 цепи коммутации для самотестирования

nonphixi0n committed:
Внес исправления с функцией DisconnectPE

nonphixi0n committed:
Мелкие правки

asytyi committed:
Удалён мёртвый дефайн

asytyi committed:
Исправлено расположение регистров

asytyi committed:
Исправлена логика записи в выходной регистр + удалён мёртвый код

asytyi committed:
Оптимизирована логика проверки одиночных реле

asytyi committed:
Исправлен баг с состояниями при проверке замкнутых реле

asytyi committed:
Удалён дублирующийся код. Исправлено форматирование

asytyi committed:
Задержка коммутации перенесена на нижний уровень, исправлена логика обработки коммутации None

asytyi committed:
Исправлена кодировка на UTF8

asytyi committed:
Исправлена очерёдность записи в цепочку регистров

asytyi reviewed (commented):
В SelfTest добавлены массивы для тестирования на залипание InputRelay, но не добавлены для MCRelay

asytyi committed:
Добавлено тестирование MCRelay на залипание

IvanovDmtr committed:
Удалено управление линией OE сдвиговых регистров, т.к. это бессмысленно в этом месте кода

IvanovDmtr committed:
Доработана таблица коммутации под платы реле новой версии

IvanovDmtr committed:
Контроль за сигналом безопасности перенесен в общий цикл
Добавлена настраеваемая задержка отключения коммутации при срабатывании безопасности
Установка State и SubState разделены на две функции

IvanovDmtr committed:
Добавлена библиотека взаимодействия с PMXU и инициализация протокола CAN Master
